util/whitelist/debugger_whitelist.go

```diff
@@ -0,0 +1,156 @@
+	logrus.Info("whiteListURL: ", whiteListURL, ", proxyCount: ", proxyCount)
+}
+
+// IsIPValid checks if the debugger IP is in the whitelist through the whitelist backend
+func IsIPValid(ip string) (bool, error) {
+	cacheKey := "whitelist-ip-" + ip
+	cacheValue, found := whiteListCache.Get(cacheKey)
+	logrus.Debug("whitelist IP cache Get ip: ", ip, ", found: ", found, ", cacheValue: ", cacheValue)
+	if found {
+		return cacheValue.(bool), nil
+	}
+
+	params := url.Values{}
+	getDebuggerURL, err := url.Parse(whiteListURL + "/api/get_debugger")
+	logrus.Debug("getDebuggerURL: ", getDebuggerURL)
+	if err != nil {
+		return false, err
+	}
+	params.Set("accept", "application/json")
+	params.Set("ip", ip)
+	getDebuggerURL.RawQuery = params.Encode()
+	urlPath := getDebuggerURL.String()
+	resp, err := http.Get(urlPath)
+	if err != nil {
+		logrus.WithError(err).Errorf("Fail to get url (%v)", urlPath)
+		return false, err
+	}
+	defer resp.Body.Close()
+	body, _ := ioutil.ReadAll(resp.Body)
+	var data map[string]interface{}
+	err = json.Unmarshal(body, &data)
+	if err != nil {
+		return false, err
+	}
+
+	debuggerList, ok := data["debugger"]
+	isValid := (ok == true) && (debuggerList != nil)
+	whiteListCache.Set(cacheKey, isValid, cache.DefaultExpiration)
+	return isValid, nil
+}
+
+// GetInvalidIPErrorMsg return `Method not supported` (code=-32004) if invalid. Reference:
+// https://github.com/ethereum/EIPs/blob/master/EIPS/eip-1474.md
+func GetInvalidIPErrorMsg(reqByte []byte) ([]byte, error) {
+	var reqs []ReqBody // for batch request
+	var req ReqBody    // for single request
+	if isBatchRequest(reqByte, &reqs) {
+		var resps []InvalidIPRespBody
+		for _, req := range reqs {
+			var resp InvalidIPRespBody
+			resp.ID = req.ID
+			resp.JsonRPC = req.JsonRPC
+			resp.Method = req.Method
+			resp.Error.Code = -32004
+			resp.Error.Message = "Operation not permitted"
+			resps = append(resps, resp)
+		}
+		return json.Marshal(resps)
+	} else if isSingleRequest(reqByte, &req) {
+		var resp InvalidIPRespBody
+		resp.ID = req.ID
+		resp.JsonRPC = req.JsonRPC
+		resp.Method = req.Method
+		resp.Error.Code = -32004
+		resp.Error.Message = "Operation not permitted"
+		return json.Marshal(resp)
+	}
+	logrus.Debug("invalid request reqByte: ", string(reqByte))
+	return []byte("invalid request " + string(reqByte)), nil
+}
+
+type ReqBody struct {
+	ID      interface{} `json:"id,omitempty"`
+	JsonRPC string      `json:"jsonrpc,omitempty"`
+	Method  string      `json:"method,omitempty"`
+}
+
+type InvalidIPRespBody struct {
+	ID      interface{}     `json:"id,omitempty"`
+	JsonRPC string          `json:"jsonrpc,omitempty"`
+	Method  string          `json:"method,omitempty"`
+	Error   InvalidIPErrMsg `json:"error,omitempty"`
+}
+
+type InvalidIPErrMsg struct {
+	Code    int64  `json:"code,omitempty"`
+	Message string `json:"message,omitempty"`
+}
+
+func isBatchRequest(reqByte []byte, reqs *[]ReqBody) bool {
+	err := json.Unmarshal(reqByte, reqs)
```

Comment:
Is it necessary to limit `reqs` length here? Or is payload size limited somewhere?

Reply:
Probably better to limit size before we read it into memory, maybe you could test it with a large payload.

Reply:
Not sure if golang http already limits this. If not, we could do something like this:

```go
r.Body = http.MaxBytesReader(w, r.Body, MaxFileSize) 
defer r.Body.Close()
body, err := ioutil.ReadAll(r.Body)
// handle error ...
```